Add Text Fragments article (#2503)
GoogleChrome/web.dev: src/site/content/en/blog/text-fragments/index.md

Like the other parameters, they, too, need to be percent-encoded.
<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=the-,text,-parameter"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=the-,text,-parameter</strong></code></a>.
To allow the parser to clearly identify the `prefix-` and the `-suffix`,
they need to be separated from the `textStart` and the optional `textEnd` with a dash&nbsp;`-`.

<figure class="w-figure">
  <img src="correct-text.png" alt="" class="w-screenshot">
  <figcaption class="w-figcaption">Text Fragment matching at the correct occurrence of <code>Text</code>.</figcaption>
</figure>

### The full syntax

The full syntax of Text Fragments is shown below (square brackets indicate an optional parameter).
The values for all parameters need to be percent-encoded.
This is especially important for the dash `-`, ampersand `&`, and comma `,` characters,
so they are not being interpreted as part of the text directive syntax.

```bash
#:~:text=[prefix-,]textStart[,textEnd][,-suffix]
```

> Each of `prefix-`, `textStart`, `textEnd`, and `-suffix` will only match text within a single

Reviewer: it's implied but might be useful to make explicit that a full `textStart,textEnd` range _can_ span multiple blocks.

Author: Added.

uninterrupted block-level element:

```html
  <div>The<div> </div>quick brown fox</div>
  <div>jumped over the lazy dog</div>
```

It does, however, match in this example:

```html
  <div>The quick brown fox</div>
  <div>jumped over the lazy dog</div>
```

### Multiple text fragments in one URL

Note that multiple text fragments can appear on one URL.
The particular text fragments need to be separated by an ampersand character `&`.
Here is an example link with three text fragments:
<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=Text%20URL%20Fragments&text=text,-parameter&text=:~:text=On%20islands,%20birds%20can%20contribute%20as%20much%20as%2060%25%20of%20a%20cat's%20diet"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=Text%20URL%20Fragments&text=text,-parameter&text=:~:text=On%20islands,%20birds%20can%20contribute%20as%20much%20as%2060%25%20of%20a%20cat's%20diet<strong></code></a>.

### Mixing element and text fragments

Traditional element fragments can be combined with text fragments.
It is perfectly fine to have both in the same URL, for example,

> to provide a meaningful fallback for browsers that do not support Text Fragments yet.

Reviewer: This doesn't really work for browsers that don't support text fragments, they won't know to strip out the `:~:` part so the browser won't find any fragment.

The more useful case here is if text on the page changes so the text fragment no longer matches, a conforming browser will then fall back to the element id which might be the nearest section header, for example.

Author: Changed to what you suggested.

User agent instructions are also called *directives*.
In the concrete case, `text=` is therefore called a *text directive*.

### Feature detectability

Support for Text Fragments is feature-detectable
by checking for the existence of the `fragmentDirective` in the `Location` interface.
Unlike the other properties like `hash` that I have mentioned above,
`fragmentDirective` does not return any value and is read-only.
The fragment directive is a mechanism for URLs to specify instructions
directed to the browser rather than the document.
It is meant to avoid direct interaction with author script,
so that future user agent instructions can be added without fear
of introducing breaking changes to existing content.
One potential example of such future additions could be translation hints.

```js
if ('fragmentDirective' in window.location) {
  // Text Fragments is supported.
}
```

### Polyfillability

The Text Fragments feature can be polyfilled to some extent.
There is [work in progress](https://github.com/tomayac/text-fragments-polyfill/)
to create an extension for browsers that do not support Text Fragments natively
where the functionality is implemented in JavaScript.

> ### Security

Reviewer: It should be noted somewhere that the text fragment is only apply to the main frame - they won't search for text inside an iframe and an iframe navigation won't invoke a text fragment.

For background, this was restricted to reduce the attack surface - we don't have any known specific attacks and this is something we could consider loosening in the future but we thought the reduced attack surface would simplify things initially and we didn't have a strong use case.

Author: Added.

Now imagine there existed a long human resources document "what to do if you suffer from…"
and then a list of conditions like "burn out", "anxiety", etc.
I could place a tracking pixel next to each item on the list.
If I then determine that loading the document temporally co-occurs
with the loading of the tracking pixel next to, say, the "burn out" item,
I can then, as the intranet admin, determine that an employee has clicked through
on a text fragment link with `:~:text=burn%20out`
that the employee may have assumed was confidential and not visible to anyone.
Since this example is somewhat contrived
and its exploitation requires very specific preconditions to be met,
on the Chrome team we have decided to implement scroll on navigation,
which is one of the preconditions.
Other user agents may decide to show a manual scroll UI element instead.
For sites that still wish to opt-out, we have proposed a
[Document Policy](https://github.com/w3c/webappsec-feature-policy/blob/master/document-policy-explainer.md)
header value that they can send, so user agents will not process Text Fragment URLs.

```bash

> Document-Policy: force-load-at-top

Reviewer: We're experimenting with this using an origin trial in M83 (applying this policy but via the origin trial, rather than document policy which is as-yet unshipped). This might be a good place to mention the trial @nickburris

Reviewer 2: Agreed, let's link to the origin trial page here. I'm just waiting for the origin trials team to start the trial, I should have a link for you this week.

Author: Added.

---
title: |

>   Link to where no human has linked to before: Text Fragments

Author: The full frontal nerdity it is :-)

was meant to be linked to by the author of the blog post.

But what about the sections I actually may have wanted to link to in real life?
Unfortunately there is not a single red box in the entire blog post body.
Say I wanted to deep link to the *ECMAScript Modules in Web Workers* heading.
As you can see in the screenshot below, the `<h1>` in question does not have an `id` attribute.
Without a fragment identifier, there is no way how I could link to this heading.
This is the problem that Text Fragments solve.

<figure class="w-figure">
  <img src="id-missing.png" alt="" class="w-screenshot" width="400">
  <figcaption class="w-figcaption">Dev Tools showing a heading without an <code>`id`</code>.</figcaption>
</figure>

## Text Fragments

The [Text Fragments](https://wicg.github.io/ScrollToTextFragment/) proposal adds support
for specifying a text snippet in the URL fragment.
When navigating to a URL with such a text fragment, the user agent can quickly emphasize
and/or bring it to the user's attention.

### `textStart`

In its simplest form, the syntax of Text Fragments is as follows:
The hash symbol `#` followed by `:~:text=` and finally `textStart`, which represents the

> [percent-encoded](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)

Author: In this case I disagree. The term "percent-encoded" appears in the RFC https://tools.ietf.org/html/rfc3986.

text I want to link to.

```bash
#:~:text=textStart
```

Taking up the example from above where I wanted to place a deep link to the
*ECMAScript Modules in Web Workers* heading, the URL in this case would be

> <a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=ECMAScript%20Modules%20in%20Web%20Workers"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=ECMAScript%20Modules%20in%20Web%20Workers</strong></code></a>

Author: Yeah, I wish I could inline-style it. @kaycebasques, is it OK to add inline styles like `<em style="color: red;">`?

> <a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=ECMAScript%20Modules%20in%20Web%20Workers"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=ECMAScript%20Modules%20in%20Web%20Workers</strong></code></a>

Reviewer 3: I would add "See it for yourself" to the next paragraph and move the link to it. That would make the bolding easier to see and the intent of the link clearer. Do the same at line 58.

> <a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=ECMAScript%20Modules%20in%20Web%20Workers"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=ECMAScript%20Modules%20in%20Web%20Workers</strong></code></a>

Author: I have solved this now by using `<mark class="highlight-line highlight-line-active">`.

The answer is that in this release Text Fragments were introduced:

<figure class="w-figure">
  <img src="text-fragments.png" alt="Blog post text: Text URL Fragments.
    Users or authors can now link to a specific portion of a page
    using a text fragment provided in a URL.
    When the page is loaded, the browser highlights the text and scrolls the fragment into view.
    For example, the URL below loads a wiki page for 'Cat'
    and scrolls to the content listed in the `text` parameter." class="w-screenshot" width="400">
  <figcaption class="w-figcaption">Text Fragments announcement blog post excerpt.</figcaption>
</figure>

If I wanted to link to the word `text`
that is written in a green code font at the bottom of the screenshot above,
I would set `textStart` to `text`.
Since the word `text` is, well, only one word, there cannot be a `textEnd`,
but even just the blog post excerpt alone contains four times the word `text`.
What now?
The URL
<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=text"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=text</strong></code></a>

> matches at the first occurrence of the word `Text`.

Author: The important part here is that "text" appears four times in this section alone, but I want to refer explicitly to the green code font one, not the other three.

> matches at the first occurrence of the word `Text`.

Reviewer 3: The way to handle this is to start the section with something like "Notice the green text in the image above." Then explain what you want readers to know about it.

> matches at the first occurrence of the word `Text`.

Author: Rephrased now.

{% Aside %}
  Note that the Text Fragment matching is case-insensitive.
{% endAside %}

<figure class="w-figure">
  <img src="first-text.png" alt="" class="w-screenshot" width="400">
  <figcaption class="w-figcaption">Text Fragment matching at the first occurrence of <code>Text</code>.</figcaption>
</figure>

Luckily there is a solution.
In cases like this, I can specify a `prefix​-` and a `-suffix`.
The word before the green code font `text` is `the`, and the word after is `parameter`.
No other of the three occurrences of the word `text` has the same surrounding words.
Armed with this knowledge, I can tweak the previous URL and add the `prefix-` and the `-suffix`.
Like the other parameters, they, too, need to be percent-encoded.
<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=the-,text,-parameter"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=the-,text,-parameter</strong></code></a>.
To allow the parser to clearly identify the `prefix-` and the `-suffix`,

> they need to be separated from the `textStart` and the optional `textEnd` with a dash&nbsp;`-`.

Author: `prefix` and `suffix` can contain multiple words, yes.

> they need to be separated from the `textStart` and the optional `textEnd` with a dash&nbsp;`-`.

Reviewer 3: You should mention that somewhere.

> they need to be separated from the `textStart` and the optional `textEnd` with a dash&nbsp;`-`.

Author: Added.

It is perfectly fine to have both in the same URL, for example,
to provide a meaningful fallback in case the original text on the page changes,
so the text fragment does not match anymore.
The URL
<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#HTML1:~:text=Give%20us%20feedback%20in%20our%20Product%20Forums."><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#HTML1:~:text=Give%20us%20feedback%20in%20our%20Product%20Forums.</strong></code></a>
linking to the *Give us feedback in our
[Product Forums](http://support.google.com/bin/static.py?hl=en&page=portal_groups.cs)* section
contains both an element fragment (`HTML1`), as well as a text fragment
(`text=Give%20us%20feedback%20in%20our%20Product%20Forums.`):

<figure class="w-figure">
  <img src="text-feedback.png" alt="" class="w-screenshot" width="237">
  <figcaption class="w-figcaption">Linking with both element fragment and text fragment.</figcaption>
</figure>

### The fragment directive

So far I have only quickly glanced over one important addition without actually explaining it:
the fragment directive `:~:`.

> To avoid compatibility issues with the usage of existing URL element fragments as shown above,

Author: The `#HTML1` part.

the [Text Fragments specification](https://wicg.github.io/ScrollToTextFragment/)
introduces the fragment directive.
The fragment directive is a portion of the URL fragment delimited by the code sequence `:~:`.
It is reserved for user agent instructions, such as `text=`,
and is stripped from the URL during loading so that author scripts cannot directly interact with it.
User agent instructions are also called *directives*.
In the concrete case, `text=` is therefore called a *text directive*.

> ### Feature detectability

Reviewer 3: ```suggestion
### Feature detection
```

> Support for Text Fragments is feature-detectable
> by checking for the existence of the `fragmentDirective` in the `Location` interface.
> Unlike the other properties like `hash` that I have mentioned above,
> `fragmentDirective` does not return any value and is read-only.

Reviewer 3: ```suggestion
To detect support, test for the read only `fragmentDirective` property on `window.location`.
```

There is early [work in progress](https://github.com/tomayac/text-fragments-polyfill/)
to create an extension for browsers that do not support Text Fragments natively
where the functionality is implemented in JavaScript.

### Security

Text fragment directives are invoked only on full (non-same-page) navigations
that are the result of a
[user activation](https://html.spec.whatwg.org/multipage/interaction.html#tracking-user-activation).
Additionally, navigations originating from a different origin than the destination
will require the navigation to take place in a
[`noopener`](https://html.spec.whatwg.org/multipage/links.html#link-type-noopener) context,
such that the destination page is known to be sufficiently isolated.
Text fragment directives are only applied to the main frame.
This means that text will not be searched inside iframes,
and iframe navigation will not invoke a text fragment.

### Privacy

It is important that implementations of the Text Fragments specification
do not leak whether a text fragment was found on a page or not.
While element fragments are fully under the control of the original page author
(remember how there was no way to link to the section

> I wanted to link to before in my running example),

Reviewer 3: What?

Author: There was no element identifier `id="something"` in the blog post, so I couldn't link.

Reviewer 3: Then that's how you should say it. It's easier to understand. Also I'd put this after the current sentence.

Author: I have rephrased.

text fragments can be created by anyone.

Imagine I ran an evil ad network `evil-ads.example.com`.

> Further imagine that in one of my ad iframes I opened a hidden cross-origin iframe

Reviewer 3: ```suggestion
Further imagine that in one of my ad iframes I opened a hidden cross-origin iframe link
```

Since a naive Text Fragments implementation might decide
that a successful match should cause a focus switch,
on `evil-ads.example.com` I could listen for the `blur` event and thus know when a match occurred.
In Chrome, we have implemented Text Fragments in such a way that the above scenario cannot happen.

Another attack might be to exploit network traffic based on scroll position.
Assume I had access to network traffic logs of my victim, like as the admin of a company intranet.
Now imagine there existed a long human resources document "what to do if you suffer from…"
and then a list of conditions like "burn out", "anxiety", etc.
I could place a tracking pixel next to each item on the list.
If I then determine that loading the document temporally co-occurs
with the loading of the tracking pixel next to, say, the "burn out" item,
I can then, as the intranet admin, determine that an employee has clicked through
on a text fragment link with `:~:text=burn%20out`
that the employee may have assumed was confidential and not visible to anyone.
Since this example is somewhat contrived
and its exploitation requires very specific preconditions to be met,

> on the Chrome team we have decided to implement scroll on navigation,

Reviewer 3: ```suggestion
the Chrome team decided to implement scroll on navigation,
```

> which is one of the preconditions.

Reviewer 3: I do't understand why scroll navigation makes text fragments safer.

Author: The opposite. We believe the risk scroll navigation adds is manageable.

Reviewer 3: Your wording doesn't convey that.

Author: I have rephrased.

</figure>

Luckily there is a solution.
In cases like this, I can specify a `prefix​-` and a `-suffix`.
The word before the green code font "text" is "the", and the word after is "parameter".
None of the other three occurrences of the word "text" has the same surrounding words.
Armed with this knowledge, I can tweak the previous URL and add the `prefix-` and the `-suffix`.
Like the other parameters, they, too, need to be percent-encoded.
<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=the-,text,-parameter"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=the-,text,-parameter</strong></code></a>.
To allow the parser to clearly identify the `prefix-` and the `-suffix`,
they need to be separated from the `textStart` and the optional `textEnd` with a dash&nbsp;`-`.

<figure class="w-figure">
  <img src="correct-text.png" alt="" class="w-screenshot" width="400">
  <figcaption class="w-figcaption">Text Fragment matching at the correct occurrence of <code>Text</code>.</figcaption>

> </figure>

Reviewer 3: @robdodson Is there a way to make screenshots take the whole width. What he's illustrating is difficult to see. https://deploy-preview-2503--web-dev-staging.netlify.app/text-fragments/

Reviewer 4: you can use `width="100%"` instead of `width="400"`

Author: I have made these three screenshots wider, since their text is important.

  el.style.border = 'solid 2px red';
});
```

I can place a deep link to any element highlighted with a red box thanks to the
[fragment identifier](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Identifying_resources_on_the_Web#Fragment)
which I then use in the [hash](https://developer.mozilla.org/en-US/docs/Web/API/URL/hash)
of the page's URL.
Assuming I wanted to deep link to the *Give us feedback in our
[Product Forums](http://support.google.com/bin/static.py?hl=en&page=portal_groups.cs)*
box in the aside, I could do so by handcrafting the URL
<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#HTML1"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<mark class="highlight-line highlight-line-active">#HTML1</mark></code></a>.
As you can see in the Elements panel of the Developer Tools, the element in question
has an `id` attribute with the value `HTML1`.

<figure class="w-figure">
  <img src="id-html1.png" alt="" class="w-screenshot" width="400">
  <figcaption class="w-figcaption">Dev Tools showing the <code>id</code> of an element.</figcaption>

> </figure>

Reviewer 5: This image is rather small on account of `width="400"`

![image](https://user-images.githubusercontent.com/4713486/81128962-9bca1580-8ef7-11ea-8d43-98111abad6c3.png)

Author: Now 600 :-)

Notice the `hash` property with the value `#HTML1`.

```js/3
new URL('https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#HTML1');
/* Creates a new `URL` object
URL {
  hash: "#HTML1"
  host: "blog.chromium.org"
  hostname: "blog.chromium.org"
  href: "https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#HTML1"
  origin: "https://blog.chromium.org"
  password: ""
  pathname: "/2019/12/chrome-80-content-indexing-es-modules.html"
  port: ""
  protocol: "https:"
  search: ""
  searchParams: URLSearchParams {}
  username: ""
}
*/
```

The fact though that I had to open the Developer Tools to find the `id`
of an element speaks volumes about the probability this particular section of the page
was meant to be linked to by the author of the blog post.

What if I want to link to something without an `id`?
Say I want to link to the *ECMAScript Modules in Web Workers* heading.
As you can see in the screenshot below, the `<h1>` in question does not have an `id` attribute,
meaning there is no way I can link to this heading.
This is the problem that Text Fragments solve 🎉.

<figure class="w-figure">

>   <img src="id-missing.png" alt="" class="w-screenshot" width="400">

Reviewer 5: This image is also small on account of `width="400"`

![image](https://user-images.githubusercontent.com/4713486/81129002-b8fee400-8ef7-11ea-9a9e-a8e58037c24c.png)

>   <figcaption class="w-figcaption">Dev Tools showing a heading without an <code>`id`</code>.</figcaption>

Reviewer 5: ```suggestion
  <figcaption class="w-figcaption">Dev Tools showing a heading without an <code>id</code>.</figcaption>
```

</figure>

## Text Fragments

The [Text Fragments](https://wicg.github.io/ScrollToTextFragment/) proposal adds support
for specifying a text snippet in the URL hash.
When navigating to a URL with such a text fragment, the user agent can emphasize

> and/or bring it to the user's attention.
> 
> ### `textStart`

Reviewer 5: ```suggestion
and/or bring it to the user's attention.

### Browser compatibility

Text Fragments are supported in version 80 and beyond of Chromium-based browsers.
At the time of writing, Safari and Firefox have not publicly signaled an intent to implement
the feature. See [Related links](#related-links) for links to the Safari and Firefox discussions.

### `textStart`
```

In its simplest form, the syntax of Text Fragments is as follows:
The hash symbol `#` followed by `:~:text=` and finally `textStart`, which represents the
[percent-encoded](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)
text I want to link to.

```bash
#:~:text=textStart
```

For example, say that I want to link to the
*ECMAScript Modules in Web Workers* heading in the [blog post announcing features in Chrome 80](https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html), the URL in this case would be:

<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=ECMAScript%20Modules%20in%20Web%20Workers"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<mark class="highlight-line highlight-line-active">#:~:text=ECMAScript%20Modules%20in%20Web%20Workers</mark></code></a>

The text fragment is emphasized <mark class="highlight-line highlight-line-active">like this</mark>.
If you click the link in a supporting browser like Chrome, the text fragment is
highlighted and scrolls into view:

<figure class="w-figure">

>   <img src="syntax-simple.png" alt="" class="w-screenshot" width="400">

Reviewer 5: Also small but perhaps this one was intentional

![image](https://user-images.githubusercontent.com/4713486/81129043-e186de00-8ef7-11ea-961e-b83ac16e3d77.png)

Author: Intentional indeed. Just the heading is important.

---
title: |
  Boldly link where no one has linked before: Text Fragments
subhead: |
  Text Fragments let you specify a text snippet in the URL fragment.
  When navigating to a URL with such a text fragment, the browser can emphasize
  and/or bring it to the user's attention.
authors:
  - thomassteiner

> date: 2020-04-03

Author: ```suggestion
date: 2020-05-06
```

They are the result of running the following snippet in DevTools.
It highlights all elements that have an `id` attribute.

```js
document.querySelectorAll('[id]').forEach((el) => {
  el.style.border = 'solid 2px red';
});
```

I can place a deep link to any element highlighted with a red box thanks to the
[fragment identifier](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Identifying_resources_on_the_Web#Fragment)
which I then use in the [hash](https://developer.mozilla.org/en-US/docs/Web/API/URL/hash)
of the page's URL.
Assuming I wanted to deep link to the *Give us feedback in our
[Product Forums](http://support.google.com/bin/static.py?hl=en&page=portal_groups.cs)*
box in the aside, I could do so by handcrafting the URL
<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#HTML1"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<mark class="highlight-line highlight-line-active">#HTML1</mark></code></a>.
As you can see in the Elements panel of the Developer Tools, the element in question
has an `id` attribute with the value `HTML1`.

<figure class="w-figure">

>   <img src="id-html1.png" alt="" class="w-screenshot" width="400">

Author: ```suggestion
  <img src="id-html1.png" alt="" class="w-screenshot" width="600">
```

Notice the `hash` property with the value `#HTML1`.

```js/3
new URL('https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#HTML1');
/* Creates a new `URL` object
URL {
  hash: "#HTML1"
  host: "blog.chromium.org"
  hostname: "blog.chromium.org"
  href: "https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#HTML1"
  origin: "https://blog.chromium.org"
  password: ""
  pathname: "/2019/12/chrome-80-content-indexing-es-modules.html"
  port: ""
  protocol: "https:"
  search: ""
  searchParams: URLSearchParams {}
  username: ""
}
*/
```

The fact though that I had to open the Developer Tools to find the `id`
of an element speaks volumes about the probability this particular section of the page
was meant to be linked to by the author of the blog post.

What if I want to link to something without an `id`?
Say I want to link to the *ECMAScript Modules in Web Workers* heading.
As you can see in the screenshot below, the `<h1>` in question does not have an `id` attribute,
meaning there is no way I can link to this heading.
This is the problem that Text Fragments solve 🎉.

<figure class="w-figure">

>   <img src="id-missing.png" alt="" class="w-screenshot" width="400">

Author: ```suggestion
  <img src="id-missing.png" alt="" class="w-screenshot" width="600">
```

Now 600 :-)

  <figcaption class="w-figcaption">Dev Tools showing a heading without an <code>id</code>.</figcaption>
</figure>

## Text Fragments

The [Text Fragments](https://wicg.github.io/ScrollToTextFragment/) proposal adds support
for specifying a text snippet in the URL hash.
When navigating to a URL with such a text fragment, the user agent can emphasize
and/or bring it to the user's attention.

### Browser compatibility

> Text Fragments are supported in version 80 and beyond of Chromium-based browsers.

Author: ```suggestion
The Text Fragments feature is supported in version 80 and beyond of Chromium-based browsers.
```

In Chrome, we have implemented Text Fragments in such a way that the above scenario cannot happen.

Another attack might be to exploit network traffic based on scroll position.
Assume I had access to network traffic logs of my victim, like as the admin of a company intranet.
Now imagine there existed a long human resources document *What to Do If You Suffer From…*
and then a list of conditions like *burn out*, *anxiety*, etc.
I could place a tracking pixel next to each item on the list.
If I then determine that loading the document temporally co-occurs
with the loading of the tracking pixel next to, say, the *burn out* item,
I can then, as the intranet admin, determine that an employee has clicked through
on a text fragment link with `:~:text=burn%20out`
that the employee may have assumed was confidential and not visible to anyone.
Since this example is somewhat contrived to begin with
and since its exploitation requires *very* specific preconditions to be met,
the Chrome security team evaluated the risk of implementing scroll on navigation to be manageable.
Other user agents may decide to show a manual scroll UI element instead.

> For sites that still wish to opt-out, we have proposed a

Reviewer 3: I hate to throw a monkey wrench into this at the 11th hour. It looks like the opt-out code laded back in mid-April. No one told us or even bothered to [close the ticket until last Thursday](https://bugs.chromium.org/p/chromium/issues/detail?id=1047900). (Here's a link showing that this [merged to 83](https://storage.googleapis.com/chromium-find-releases-static/942.html#942995187c607e433af35e114a325c4ca2576d37) some time ago.)

cc: @kaycebasques, @petele

Reviewer: Sorry about the confusion (I've reopened that bug). The thing that landed and was merged back to M83 was the origin trial - the DocumentPolicy remains unshiped. So I think this text is still correct as-is.
